CIP-0180? | Block Producer Identification (#1157)
cardano-foundation/CIPs: CIP-0180/README.md

---
CIP: 180
Title: Block Producer Identification
Category: Ledger
Status: Proposed
Authors:
    - Samuel Leathers <samuel.leathers@iohk.io>
    - Adam Dean <adam@crypto2099.io>
Implementors: []
Discussions:
    - Original PR: https://github.com/cardano-foundation/CIPs/pull/1157
Created: 2026-03-04
License: CC-BY-4.0
---

## Abstract

> This CIP proposes a modification to Cardano blocks to be introduced in a
> future ledger era Dijkstra:

Reviewer: This is unlikely to go into Dijkstra, since we are too close to the deadline and there is still a lot of work that needs to be done
```suggestion
This CIP proposes a modification to Cardano blocks to be introduced in a
future ledger era:
```

Author: If we don't do it for Dijkstra, we might as well just not do it. The whole reason for this is to get ahead of the mess we're going to have of alternative node implementation reporting their versions in non-standard, hard to read ways. As soon as we have the first alternative node making blocks (which Dingo already is on testnets), there's really no good reason to go forward with this. IMHO, it's either we put this in Dijkstra so it's not a confusing mess, or its already a confusing mess and there's no good reason to change anything.

Without this capability, the community lacks visibility into critical ecosystem
health metrics derived from on-chain block production data, especially as the
ecosystem diversifies with multiple block-producing implementations.

## Specification

As a Ledger category CIP per CIP-0084<sup>3</sup>, the specification below is
intended to be sufficiently detailed for inclusion in a formal ledger
specification. The complete CDDL specification for these changes will be added
to Cardano Blueprint<sup>2</sup> as the authoritative reference for all
implementations.

### Block Structure Modifications

This proposal adds a new optional element to the block body for block producer
identification, wrapped in an extensible `informational_data_set`:

**Modified Block Definition**:

```cddl
; Block structure - 6th element added for producer identification
block =
  [ header
  , transaction_bodies       : [* transaction_body]
  , transaction_witness_sets : [* transaction_witness_set]
  , auxiliary_data_set       : {* transaction_index => auxiliary_data}
  , invalid_transactions     : [* transaction_index]
  , informational_data_set / nil  ; NEW - optional

>   ]

Reviewer: [CIP-176](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0176) has already been merged and mostly implemented.
So this block structure is no longer accurate. It is better not to assume what will be the structure of the block body when this fields will be added

```suggestion
block = [header, block_body]

block_body =
  [ ... current block body fields
  , producer_agent : tstr .size (1..32) / nil  ; NEW - optional
  ]
```

informational_data_set =
  [ producer_agent : tstr .size (1..32) / nil ]
```

The `informational_data_set` is a CBOR array designed for future extensibility:
additional fields can be appended in future eras without requiring a new era

> boundary for the outer block structure.

Reviewer: This is an incorrect statement. Any change to the serialization requires a new era. Therefore, I'd suggest to not be clever about it and not assume there more fields like this that will be added to the block body

```suggestion
```

### Producer Agent Field

Add a new optional text field inside `informational_data_set` for node

> identification:

Reviewer: ```suggestion
```

```

**Field Specifications**:

- **Type**: UTF-8 text string or `nil`
- **Size**: 1-32 bytes when present as a string (matches Ethereum graffiti
  limit<sup>6</sup>); use `nil` to opt out
- **Format**: Recommended `<implementation>/<version>[+<features>]`
- **Examples**:
  - Mainnet: `"cardano-node/10.7.0"`, `"amaru/1.0.0"`, `"dingo/0.5.2"`
  - Testnets: `"cardano-node/11.0.0+leios"`, `"amaru/1.1.0+peras"`,
    `"dingo/0.6.0+leios+peras"`
- **Validation**: Nodes **MUST** accept any valid UTF-8 content within size limit
- **Optional**: Block producers **MAY** use `nil` for operational privacy

**Content Guidelines** (following RFC 9110, Section 10.1.5<sup>1</sup>):
- **SHOULD** use format `<implementation>/<version>[+<feature>...]`
- **MAY** append build metadata using `+` separator (e.g., "+leios",
  "+peras+mithril")
- Build metadata is particularly useful on testnets for identifying experimental
  features
- **MUST NOT** include advertising, pool metadata, or ticker symbols
- **SHOULD NOT** include pool-specific information that enables fingerprinting

### Validation Rules

**Consensus Layer**:
- Nodes **MUST** accept blocks with `informational_data_set` as `nil`, as a

>   1-element array containing `nil`, or as a 1-element array containing any

Reviewer 2: That is weird? Why not simply `nil`? Why wrap it in an array?

  well-formed UTF-8 text string within the 32-byte limit
- Nodes **MUST NOT** reject blocks based on `producer_agent` content

> - This field is purely informational and does not affect block validation

Reviewer 2: I suppose that this isn't true? What if the field is invalid? (e.g. contains non-UTF-8 characters, or is too long, or is not a string). And what about the block body hash; the next paragraph suggests that it's not ignored; and thus not purely informational.

Reviewer: I sort of agree that these statements are conflicting.
I says that it must be well-formed utf-8, while the other says that block can not be rejected on its basis.
I suspect the intent was that for a structurally valid `producer_agent` it can be used  to reject blocks on teh field's contents. By "structurally", meaning to one that adheres to CDDL and it doesn't take the block_body over the max_block_body size limit.

a future ledger era Dijkstra.

**Breaking Changes**:

- Block structure changes from 5 to 6 elements
- Block body hash computation changes to include new 6th element
- Nodes running pre-hard-fork software cannot deserialize post-HF blocks

**Pre-HF (Conway)**:
- Blocks **MUST** use 5-element structure
- Block body hash computed over 5 elements

**Post-HF (Dijkstra)**:
- Blocks **MUST** use 6-element structure
- Block body hash computed over 6 elements (including `informational_data_set`)
- `informational_data_set` **MAY** be `nil` or a 1-element array containing
  `producer_agent` as `nil` or a text string (1-32 bytes)

**Historical Blocks**:
- Pre-HF blocks remain valid with original 5-element structure
- No retroactive changes to historical block interpretation

## Rationale: How does this CIP achieve its goals?

### Design Decisions

**Block Body Placement**:
We chose to place the producer identification field in the block body rather
than the header because:

- **Header Performance**: Headers are on the critical path for block propagation
  - keeping them minimal is important for network performance
- **Optional Without Overhead**: A `nil` value in the block body adds only 1

>   byte, whereas header fields have higher overhead

Reviewer: No they don't have higher overhead. Same design applicable if it was included in block header.
```suggestion
- **Optional With Minimal Overhead**: A `nil` value in the block body adds only 1
  byte.
```

- **Clear Separation**: Producer identification is metadata about the block, not
  part of chain validation
- **Future Extensibility**: The `informational_data_set` wrapper allows adding

>   producer metadata fields in future without modifying the outer block structure

Reviewer: As mentioned earlier this is invalid
```suggestion
```

- **Era Transition**: Block body hash changes are acceptable in a new era as
  all nodes must upgrade anyway

**`informational_data_set` Wrapper**:
The `producer_agent` field is wrapped in an `informational_data_set` array
rather than placed directly in the block body because:

- **Future Extensibility**: Additional informational fields can be appended to
  the array in future eras without changing the outer block structure
- **Semantic Grouping**: Groups related informational metadata together cleanly
- **Minimal Overhead**: A `nil` top-level value still costs only 1 byte when not

>   used; the array itself is only a few bytes when present

Reviewer: Ditto
```suggestion
```

**Free-Form Text vs. Structured Data**:
We chose free-form text for the `producer_agent` field because:

- **Flexibility**: Allows block producers to include relevant information
  without protocol restrictions
- **Future-Proofing**: New node implementations (cardano-node, Amaru, Dingo,
  Torsten, Gerolamo, and future implementations) can participate without
  requiring protocol updates to register identifiers
- **Simplicity**: Easier to implement and maintain than structured versioning
  schemes or registry systems
- **User-Agent Precedent**: Follows successful patterns from HTTP (RFC 9110) and
  other protocols
- **Testnet Features**: Enables signaling of experimental features (Leios,
  Peras, Mithril) using build metadata notation

**1-32 Byte Range**:
The length range balances functionality with efficiency:

- **Minimum of 1**: Empty strings carry no useful information; `nil` already
  serves as the opt-out value, so a zero-length string would be redundant

> - **Maximum of 32**: Accommodates implementation name, version, and build

Reviewer 2: It still feels a bit like a waste of space. And without canonical taxonomy for the features, it'll be a mess real fast for consumers. We don't need 32 bytes to identify nodes IMO. Realistically, we'll likely have less than a dozen of nodes so one letter could suffice. If we really want it, 2 letters give us up to 676 node implementations. `cn`, `am`, `di`, `to`. For the version; only a patch version is likely relevant as the first two numbers are likely implied by the protocol version? I mean, _it could be_ at least. And that patch version should imply features already? I don't quite see the need for the extra `+...`

I like the freedom and expressiveness in that; but I also dislike breaking changes. As I already suggested during the last node diversity workshop, we _could_ make that fit within the protocol version _minor_ where we have a full 8 bytes of data. It's in the header, but, it's _already_ there. 

And we could come up with an encoding which fits in a u64 (and that still carry the minor version!). No breaking-change or new era required; and forces compactness.

Reviewer 3: I support the general idea of this CIP to have a block minted measurement way of measurements. 

As many things have been said here by others I would just tag @nfrisby because I remember him also reacting surprised on the idea to make this a whole user-agent like string, instead of just 2 bytes or existing fields in the header.

  (32 bytes)<sup>6</sup>, a proven approach
- **Denial-of-Service Prevention**: Prevents abuse through excessive data
- **Network Efficiency**: Minimal storage impact. At Cardano's rate of ~1 block
  per 20 seconds (~1,576,800 blocks/year):
  - Typical case (19-byte strings): ~30-40 MB/year
  - Worst case (32-byte strings): ~50-63 MB/year
  - Current chain growth: ~45-50 GB/year
  - Impact: ~0.05-0.14% increase
- **Human-Readable**: Unlike hash-based approaches, readable by operators and
  analysts

**Optional Field**:
Making `producer_agent` optional via `nil`:

- **Gradual Adoption**: Block producers can adopt at their own pace
- **Privacy**: Permits operators who prefer not to disclose implementation
  details for operational security
- **Flexibility**: No requirement to populate even when node software supports it
- **Minimal Overhead**: `nil` adds only 1 byte when not used

### Alternative Approaches Considered

**Block Header Placement**:
Adding the field to the block header was rejected in favor of block body
placement because:

- Headers are on the critical path for block propagation - avoiding header
  changes preserves performance

> - Block body allows cleaner optional semantics (`nil` = 1 byte)

Reviewer: Ditto. Invalid statement
```suggestion
```

forged a given block — nothing more. Nodes with no stake or low luck in a given
epoch may produce no blocks at all, making their software invisible during that
period.

## CBOR Encoding Examples

The 6th block element is encoded as either `nil` (1 byte) or a 1-element array:

```
-- With producer agent string
Diagnostic: ["cardano-node/10.7.0"]
Hex: 81                              -- array(1)
     73                              -- text(19)
     636172 64616e6f2d6e6f64652f31302e372e30  -- "cardano-node/10.7.0"

-- Producer agent omitted
Diagnostic: [nil]
Hex: 81 f6                           -- array(1), nil

-- No informational data
Diagnostic: nil
Hex: f6                              -- nil (1 byte)
```

## Path to Active

### Acceptance Criteria

- [ ] CDDL schema updates merged into Cardano Blueprint as the authoritative
  specification
- [ ] All node implementation teams have reviewed and agreed upon the
  specification (cardano-node, Amaru, Dingo, Torsten, Gerolamo)
- [ ] Implementation present in cardano-node (Haskell) released as part of
  Dijkstra hard fork
- [ ] Implementation present in Amaru (Rust)
- [ ] Implementation present in Dingo (Go)

> - [ ] Implementation present in Torsten (Rust)

Reviewer 4: ```suggestion
- [ ] Implementation present in Dugite (Rust)
```
As I understand this has node project [has been renamed](https://github.com/michaeljfazio/dugite/commit/6ceb17c98354e310900b7d93503a415d12a9776a)... cc @michaeljfazio

Reviewer 5: Correct. "Torsten" is now "Dugite"

- [ ] Implementation present in Gerolamo (TypeScript)
- [ ] The feature is activated on mainnet through the Dijkstra new era
- [ ] Chain analytics tools have been updated to parse and display producer
  agent data
- [ ] Block explorers correctly display blocks from both pre- and post-hard-fork
  eras

### Implementation Plan

This CIP requires coordination across all node implementations as part of the

> Dijkstra era hard fork. Each implementation (cardano-node, Amaru, Dingo,

Reviewer: ```suggestion
This CIP requires coordination across all node implementations as part of the
new era hard fork. Each implementation (cardano-node, Amaru, Dingo,
```
